crowbar_framework/app/models/api/node.rb

```diff
@@ -50,37 +50,126 @@ def repocheck(options = {})
       end
     end
 
-    protected
+    def initialize(name = nil)
+      @node = NodeObject.find_node_by_name name
+    end
 
-    def node_architectures
-      {}.tap do |ret|
-        NodeObject.all.each do |node|
-          arch = node.architecture
-          ret["os"] ||= []
-          ret["os"].push(arch) unless ret["os"].include?(arch)
-
-          if ceph_node?(node)
-            ret["ceph"] ||= []
-            ret["ceph"].push(arch) unless ret["ceph"].include?(arch)
-          else
-            ret["openstack"] ||= []
-            ret["openstack"].push(arch) unless ret["openstack"].include?(arch)
-          end
+    def upgraded?
+      # FIXME: check this by looking at some upgraded-ok file on the node
+      current_platform = "#{@node[:platform]}-#{@node[:platform_version]}"
+      @upgraded = current_platform == @node[:target_platform]
+    end
 
-          if pacemaker_node?(node)
-            ret["ha"] ||= []
-            ret["ha"].push(arch) unless ret["ha"].include?(arch)
-          end
-        end
+    def pre_upgrade
+      # TODO: save the global status info about this substep (we started prepare for upgrade)
+
+      # Migrate out the l3 agents and shut down pacemaker
+      script = "/usr/sbin/crowbar-pre-upgrade.sh"
+      out = @node.run_ssh_cmd(script)
+      unless out[:exit_code].zero?
+        Rails.logger.error("Executing of pre upgrade script has failed on node #{@node.name}.")
+        Rails.logger.error("Script location: #{script}")
+        Rails.logger.error("stdout: #{out[:stdout]}, stderr: #{out[:stderr]}")
+        return false
       end
+      true
     end
 
-    def ceph_node?(node)
-      node.roles.include?("ceph-config-default")
+    def os_upgrade
+      # Upgrade one node
+      # TODO: save the global status info about this substep (we started upgrade of the node)
+      ssh_status = @node.ssh_cmd("/usr/sbin/crowbar-upgrade-os.sh")
+      if ssh_status[0] != 200
+        Rails.logger.error("Executing of os upgrade script has failed on node #{@node.name}.")
+        return false
+      end
+      true
+    end
+
+    def post_upgrade
+      # FIXME: so far, we have no post-upgrade script
+      return true
+      # TODO: save the global status info about this substep (we started post upgrade stuff)
+
+      # Join the cluster: start pacemaker and run selected recipes
+      script = "/usr/sbin/crowbar-post-upgrade.sh"
```

Comment:
Yes, that's fine, we do not have post_ugrade method yet (that's part of orchestration of more nodes)